### ISSUE TITLE
Lower mAP with cache: True

### ISSUE THREAD

**Issue by alixunxing** [ultralytics/ultralytics#18938]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Did anyone try to train a detector model with cache=True?
I have trained several models with cached images but it always has a lower mAP (3% lower) compared with no cache.
Is there any relationship with the cache parameter?
And I have seen the warning "WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows."
I don't understand why it may produce non-deterministic results. Since all the images are just loaded into memory first.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @alixunxing, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it effectively.

If this is a custom training ❓ Question, please provide as much detail as possible, such as:
- The training settings and hyperparameters you used
- Dataset image examples and labels
- Log outputs or training metrics
And ensure that you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Regarding the specific behavior of the `cache` parameter and the observed impacts on mAP, this is an interesting finding. The provided warning "cache='ram' may produce non-deterministic training results" appears because caching in RAM could result in subtle variations due to memory handling, multiprocessing, or even operations executed in a non-deterministic order. For a more stable alternative, `cache='disk'` allows deterministic behavior if sufficient disk space is available.

Join the Ultralytics community where it suits you best:
- For real-time help, chat with us on [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- Share experiences on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Ensure you are using the latest `ultralytics` version to rule out any previously resolved issues. Update your environment using:

```bash
pip install -U ultralytics
```

See our [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) for compatibility, and ensure you're working in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/).

## Environments

YOLO can run smoothly on several verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests confirm correct functionality across all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

This is an automated response to assist you promptly. An Ultralytics engineer will follow up soon to provide further help! 😊

**Comment by glenn-jocher** [comment]
@alixunxing thank you for your question. The `cache=True` option, particularly `cache='ram'`, can lead to non-deterministic training results due to potential subtle differences in how images are processed when cached in RAM versus being loaded directly from disk. This can affect augmentation or preprocessing randomness, potentially causing the observed mAP differences. For consistent results, consider using `cache='disk'` as suggested in the warning. You can find more details about caching in the [documentation here](https://docs.ultralytics.com/guides/model-training-tips/#caching). Let us know if you have further questions.

**Comment by alixunxing** [comment]
Another interesting thing:

Without cache enabled, the last 10 epoch (stop mosaic augmentation) appears a decreased training loss as the following image shows.

![Image](https://github.com/user-attachments/assets/01874015-4578-443f-a1ee-73572fa931d3)

But if I turn on cache, the loss in the last 10 epochs will increase dramatically.

![Image](https://github.com/user-attachments/assets/0b95bb73-cad0-4004-85a9-a9d63c125296)

**Comment by glenn-jocher** [comment]
@alixunxing the behavior you're observing with cached training and mosaic augmentation is likely due to how cached images interact with dynamic augmentation changes. When `cache=True`, the original images are stored, but **augmentations are still applied on-the-fly** each epoch. However, the mosaic disablement in final epochs (`close_mosaic=10`) combined with cached images can sometimes create subtle interactions in the augmentation pipeline that affect loss trajectories. 

For best stability with mosaic training, we recommend either:
1. Using `cache='disk'` instead of `cache='ram'` 
2. Keeping mosaic disabled (`close_mosaic=0`) if using RAM caching

This tradeoff between speed and stability is why we include the warning about non-deterministic results with RAM caching. The team is actively working to improve caching reliability in future releases.

**Comment by Y-T-G** [comment]
@alixunxing Does the problem occur if you use this PR?

https://github.com/ultralytics/ultralytics/pull/20474

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR (#20474), authored by Y-T-G with contributions from glenn-jocher.

The update introduces automatic selection between buffer and full-dataset image sampling for mosaic augmentation, depending on your dataset's cache setting. This streamlines the training process, reduces configuration steps, and helps prevent user errors when working with different caching strategies. These improvements should lead to more consistent and efficient data augmentation with less manual intervention.

To test if these changes resolve your issue, try upgrading to the latest version with:
```
pip install -U ultralytics>=8.3.129
```
Alternatively, you can install directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
You can also wait for the next official PyPI release if you prefer.

Please let us know if this update addresses your issue, or if you experience any further problems. Thanks so much for reporting this and sharing your feedback! 🙏

**Comment by Petros626** [comment]
@alixunxing any update to the suggested pr?
